reasoner_pydantic/upgrade.py

```diff
@@ -0,0 +1,194 @@
+            ):
+                if provenance_tree:
+                    check = True
+                    for source in provenance_tree:
+                        if source["resource"] == attribute.get("attribute_source"):
+                            provenance_tree.append(
+                                provenance_tree[len(provenance_tree) - 1]
+                            )
+                            source["retrievals"] = attribute["value"]
+                            source_index = provenance_tree.index(source)
+                            for i in range(
+                                len(provenance_tree) - 1, source_index + 1, -1
+                            ):
+                                provenance_tree[i] = provenance_tree[i - 1]
+                            provenance_tree[source_index] = {
+                                "resource": attribute["value"],
+                                "resource_role": attribute["attribute_type_id"],
+                            }
+                            check = False
+                            break
+                    if check:
+                        provenance_tree.append(
+                            {
+                                "resource": attribute["value"],
+                                "resource_role": attribute["attribute_type_id"],
+                                "retrievals": [
+                                    provenance_tree[len(provenance_tree) - 1][
+                                        "resource"
+                                    ]
+                                ],
+                            }
+                        )
+                else:
+                    provenance_tree.append(
+                        {
+                            "resource": attribute["value"],
+                            "resource_role": attribute["attribute_type_id"],
+                        }
+                    )
+            else:
+                kattributes.append(attribute)
+        edge_key = hash(
+            (
+                ksubject,
+                kobject,
+                kpredicate,
+                knegated,
+                tuple(hashing_qualifiers),
+                ksource,
+            )
+        )
+        converted_edge = {
+            "subject": ksubject,
+            "predicate": kpredicate,
+            "object": kpredicate,
+            "negated": knegated,
+            "qualifiers": kqualifiers,
+            "sources": provenance_tree,
+            "attributes": kattributes,
+        }
+        return edge_key, converted_edge
+
+    # Actually do the upgrading
+    kg = old_dict["knowledge_graph"]
+    new_edges = dict()
+    edge_key_map = dict()
+    for old_key, old_edge in kg["edges"].items():
+        edge_key, converted_edge = edge_conversion(old_edge)
+        edge_key_map[old_key] = edge_key
+        new_edges[edge_key] = converted_edge
```

Comment:
Will this preserve attributes and source retrieval provenance? It seems like the edge is just replaced, so when updating the message, we might lose some information.

Reply:
Yes, we need to create the new edge better so as to preserve all provenance information.